proposals/NNNN-forward-scan-trailing-closures.md

```diff
@@ -0,0 +1,295 @@
+
+Note that this API interacts poorly with SE-0279 multiple trailing closures, because the unlabeled trailing closure "moves" backwards as additional trailing closures are provided at the call site:
+
+```swift
+// SE-0279 backward scan behavior
+BlockObserver { (operation, errors) in
+  print("finishHandler!")
+}
+
+// label finishHandler, unlabeled moves "back" to produceHandler
+
+BlockObserver { (aOperation, foundationOperation) in
+  print("produceHandler!")
+} finishHandler: { (operation, errors) in
+  print("finishHandler!")
+}
+
+// label produceHandler, unlabeled moves "back" to startHandler
+BlockObserver { aOperation in 
+  print("startHandler!")
+} produceHandler: { (aOperation, foundationOperation) in
+  print("produceHandler!")
+} finishHandler: { (operation, errors) in
+  print("finishHandler!")
+}
+```
+
+The forward scan provides a consistent unlabeled trailing closure anchor, and later (labeled) trailing closures can be tacked on:
+
+```swift
+// Proposed forward scan
+BlockObserver { aOperation in 
+  print("startHandler!") {
+}
+
+// add another
+BlockObserver { aOperation in 
+  print("startHandler!")
+} produceHandler: { (aOperation, foundationOperation) in
+  print("produceHandler!")
+}
+
+// specify everything 
+BlockObserver { aOperation in 
+  print("startHandler!")
+} produceHandler: { (aOperation, foundationOperation) in
+  print("produceHandler!")
+} finishHandler: { (operation, errors) in
+  print("finishHandler!")
+}
+
+// skip the middle one!
+BlockObserver { aOperation in 
+  print("startHandler!")
+} finishHandler: { (operation, errors) in
+  print("finishHandler!")
+}
+```
+
+The forward-scan matching rule provides more predictable results, making it easier to understand how to use this API properly. 
+
+### Workaround via overload sets
+
+APIs like the above that will be broken by the forward scan can be "fixed" by removing the default arguments, then adding additional overloads to create the same effect. For example, drop the default argument of `finishHandler` so that the heuristic will kick in to fix calls with a single unlabeled trailing closure:
+
+```swift
+init(startHandler: ((AOperation) -> Void)? = nil, produceHandler: ((AOperation, Foundation.Operation) -> Void)? = nil, finishHandler: ((AOperation, [NSError]) -> Void)) {
+    self.startHandler = startHandler
+    self.produceHandler = produceHandler
+    self.finishHandler = finishHandler
+}
+```
+
+One can then add overloads to handle other cases, e.g., the zero-argument case:
+
+```swift
+init() {
+  self.init(startHandler: nil, produceHandler: nil, finishHandler:nil)
+}
+```
+
+An investation into APIs with multiple trailing closures uncovered one SwiftUI API that will break in this manner, because both closure parameters have defaults:
```

Comment:
`investation` → `investigation`

```suggestion
An investigation into APIs with multiple trailing closures uncovered one SwiftUI API that will break in this manner, because both closure parameters have defaults:
```